docs/proposals/config_data.md

```diff
@@ -0,0 +1,262 @@
+
+### Dynamic Configuration
+
+Any long-running system has mutating specification over time. In order to facilitate this functionality,
+``ConfigData`` will be versioned and updates will automatically made available to the container (downward API).
+
+``resourceVersion`` (Found in ``ObjectMetadata``) of the ``ConfigData`` object will be updated by the server every time the object is modified.
+After an update, modifications will be made visible to the consumer container. If the consumer uses the ``Data`` pairs
+only for initialization or during starting process, A rolling-update might be necessary to update the components.
+
+It is then the consumer component's responsilibity to make use of the updated data
+once it is made visible (or perform a rolling-update on consumers of that object).
+
+### Advantages
+
+* Reusable across different components.
+* Easy distribution through the API server.
+* Ability to use ``Kubernetes`` authentication and security policies with configuration objects.
+* By leveraging the power of ``/watch`` API along with the new resource, we can change component configurations dynamically.
+* Layer of abstraction that gets rid of the global state currently used for command-line flags and environment variables.
+
+### API Resource
+
+A new resource for ``ConfigData`` will be added to the ``API``:
+
+```go
+	type ConfigData struct {
+		TypeMeta   `json:",inline"`
+		ObjectMeta `json:"metadata,omitempty"`
+
+		Data map[string]string `json:"data,omitempty"`
+	}
+
+	type ConfigDataList struct {
+		TypeMeta `json:",inline"`
+		ListMeta `json:"metadata,omitempty"`
+
+		Items []ConfigData `json:"items"`
+	}
+```
+``Registry`` interface will be added to accompany the ``ConfigData`` resource.  
+``ConfigData`` information will be stored in ``etcd`` by default.
+
+### Volume Source
+
+A new ``ConfigDataSource`` type of volume source containing the ``ConfigData`` object will be added to the ``VolumeSource`` struct in the API:
+
+```go
+	type ConfigDataVolumeSource struct {
+		ConfigDataName string `json:"configDataName"`
+	}
+```
+
+This volume source will be made available and updated by using the downward API volume plug-in (might require some generalizations).
+
+
+## Creating ConfigData
+``ConfigData`` can be created in two ways explicitly and implicitly.
+
+### Backwards-Compatibility
+In order to smooth the transaction of component configuration from other means to ``ConfigData``
+and preserve backwards-compatibility, user will be able to configure components the old way
+which is going to be transformed into a ``ConfigData`` object implicitly which can be updated and reused.
+
+For example, a ``kubelet`` is configured currently as follows:
+```bash
+kubelet --address=0.0.0.0 --port=8888 --api-servers=10.240.13.14:8998,10.240.51.5:8998 --register-node=true
+```
+
+After this proposal is integrated, when you invoke the same commands,
+a ``ConfigData`` object will be created in the API server.
```

Comment:
Depends.  However, that's a detail of how config data is used, and probably
should be a separate proposal (we are pretty much closed on the API design,
and I don't foresee any shocks when we leverage it for system components).

On Sep 14, 2015, at 7:27 PM, Andy Goldstein <notifications@github.com>
wrote:

In docs/proposals/config_data.md
<https://github.com/kubernetes/kubernetes/pull/6477#discussion_r39459716>:

> +
> +## Creating ConfigData
> +``ConfigData`` can be created in two ways explicitly and implicitly.
> +
> +### Backwards-Compatibility
> +In order to smooth the transaction of component configuration from other means to ``ConfigData``
> +and preserve backwards-compatibility, user will be able to configure components the old way
> +which is going to be transformed into a ``ConfigData`` object implicitly which can be updated and reused.
> +
> +For example, a ``kubelet`` is configured currently as follows:
> +```bash
> +kubelet --address=0.0.0.0 --port=8888 --api-servers=10.240.13.14:8998,10.240.51.5:8998 --register-node=true
> +```
> +
> +After this proposal is integrated, when you invoke the same commands,
> +a ``ConfigData`` object will be created in the API server.

I'm not too keen on having ConfigData created implicitly. In OpenShift, we
can do openshift start --write-config=/path/to/config/dir, and that will
write the configuration file(s) to the specified path. Once we've done
that, we can then start for real with openshift start
--master-config=/path/to/master-config.yaml
--node-config=/path/to/node-config.yaml. I realize this is writing to disk
and therefore isn't exactly the same scenario, but I think we could have
something like kubelet --write-config=mykubeletconfigname | kubectl create
-f - and that would create a ConfigData called mykubeletconfigname. What
does everyone else think?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/6477/files#r39459716>.